public/jira/raw/rollback/rollback.ts

```diff
@@ -0,0 +1,311 @@
+  for (const issue of issues) {
+    for (const version of issue.fields["Fix versions"]) {
+      ids.set(version.id, version);
+      names.set(version.name, version);
+    }
+  }
+  return { ids, names };
+}
+
+function getStatusesFromIssues(issues: JiraIssue[]): FieldLookup<Status> {
+  const ids = new Map<string, Status>();
+  const names = new Map<string, Status>();
+  for (const issue of issues) {
+    ids.set(issue.fields.Status.id, issue.fields.Status);
+    names.set(issue.fields.Status.name, issue.fields.Status);
+  }
+  return { ids, names };
+}
+
+export function rollbackIssues(
+  issues: JiraIssue[],
+  rollbackTime: Date = oneHourAgo
+): RolledBackJiraIssue[] {
+  const sprints = getSprintsMapsFromIssues(issues);
+  const versions = getVersionsFromIssues(issues);
+  const statuses = getStatusesFromIssues(issues);
+  return issues
+    .map((i) => rollbackIssue(i, { sprints, versions, statuses }, rollbackTime))
+    .filter((i): i is RolledBackJiraIssue => i !== undefined);
+}
+
+const oneHourAgo = new Date(Date.now() - 1000 * 60 * 60);
+
+/**
+ * @typedef {{
+ *   rolledvackTo: Date,
+ *   didNotExist: Boolean
+ * }} RolledBackMetadata
+ */
+
+/**
+ * @typedef {import("../../shared/types").JiraIssue & {rollbackMetadata: RolledBackMetadata}} RolledBackJiraIssue
+ */
+
+/**
+ * @param {import("../../shared/types").JiraIssue} issue
+ * @param {RollbackLookupData} data
+ * @param {Date} rollbackTime
+ * @returns {RolledBackJiraIssue | undefined}
+ */
+export function rollbackIssue(
+  issue: JiraIssue,
+  data: RollbackLookupData,
+  rollbackTime: Date = oneHourAgo
+): RolledBackJiraIssue | undefined {
+  const { changelog, ...rest } = issue;
+  const rolledBackIssue: RolledBackJiraIssue = {
+    ...rest,
+    rollbackMetadata: { rolledbackTo: rollbackTime },
+  };
+  // ignore old issues
+  if (parseDateISOString(issue.fields.Created) > rollbackTime) {
+    return undefined;
+    /*
+        copy.rollbackMetadata.didNotExist = true;
+        delete copy.fields;
+        // should convert to date ...
+        copy.rollbackMetadata.didNotExistBefore = issue.fields.Created;
+        return copy;*/
+  }
+
+  rolledBackIssue.fields = { ...issue.fields };
+
+  for (const { items, created } of changelog || []) {
+    // we need to go back before ...
+    if (parseDateISOString(created) < rollbackTime) {
+      break;
+    }
+    items.forEach((change) => {
+      const { field } = change;
+      const fieldName = fieldAlias[field] || field;
+      if (fields[fieldName]) {
+        Object.assign(
+          rolledBackIssue.fields,
+          fields[fieldName](
+            rolledBackIssue.fields[fieldName],
+            change,
+            fieldName,
+            data
+          )
+        );
+      } else {
+        rolledBackIssue.fields[fieldName] = change.from;
+      }
+    });
+  }
+  return rolledBackIssue;
+}
+
+/*
+
+  export function collectChangelog(observableBaseIssues: JiraIssue[], priorTime: Date) {
```

Comment:
This from the js file?

Reply:
Yup I just added placeholder types to it

Reply:
Reverted back the types